scripts/generate_matrix.py

```diff
@@ -0,0 +1,227 @@
+# !/usr/bin/env python3
+"""
+Script to generate a test matrix of Python and package versions for CI testing.
+
+This script fetches the list of supported Python versions and the versions of a
+specified package available on PyPI, determines compatible combinations based
+on package 'requires_python' metadata, and outputs a JSON matrix suitable for
+use in CI workflows.
+
+Features:
+- Generalized to work with any package (not just NumPy).
+- Caches HTTP requests to reduce network calls.
+- Prevents code duplication by using reusable functions.
+- Includes thorough doctests and documentation.
+
+Usage:
+    python generate_matrix.py [PACKAGE_NAME]
+
+Example:
+    python generate_matrix.py numpy
+
+"""
+
+import dataclasses
+import importlib.metadata
+import json
+import sys
+import urllib.error
+import urllib.request
+from functools import lru_cache
+
+# Constants for caching
+CACHE_SIZE = 128  # Adjust as needed
+
+
+@dataclasses.dataclass(frozen=True, eq=True, order=True)
+class Version:
+    major: int
+    minor: int
+    micro: int
+
+    def __str__(self) -> str:
+        return f"{self.major}.{self.minor}.{self.micro}"
+
+    @property
+    def minor_version(self) -> tuple[int, int]:
+        return self.major, self.minor
+
+    @classmethod
+    def from_str(cls, version_str: str) -> "Version":
+        if version_str.count(".") == 1:
+            version_str += ".0"
+        return cls(*map(int, version_str.strip().split(".")))
```

Comment:
This wekk, `scipy` will release `1.15.0rc1`, which we'll be using to prepare for the upcoming `scipy-stubs==1.15.0.0` release. But it looks like that'll result in a  `ValueError` here.

Reply:
True... I assumed we only wanted to test for final releases. I'm explicitly ignoring pre-releases.